File: navpipe/models/nav.py
from __future__ import annotations
import msgspec


class NavData(msgspec.Struct):
    date: str
    nav: str


class NavMeta(msgspec.Struct):
    fund_house: str
    scheme_type: str
    scheme_category: str
    scheme_code: int
    scheme_name: str
    isin_growth: str | None
    isin_div_reinvestment: str | None


class NavResponse(msgspec.Struct):
    meta: NavMeta
    data: list[NavData]


class DateRange(msgspec.Struct, frozen=True):
    start_date: str
    end_date: str

File: src/schema/nav.rs
use serde::Deserialize;

/// Single NAV entry
#[derive(Debug, Clone, Deserialize)]
pub struct NavData {
    pub date: String,
    pub nav: String,
}

/// Metadata for the fund/scheme
#[derive(Debug, Clone, Deserialize)]
pub struct NavMeta {
    pub fund_house: String,
    pub scheme_type: String,
    pub scheme_category: String,
    pub scheme_code: u32,
    pub scheme_name: String,
    pub isin_growth: Option<String>,
    pub isin_div_reinvestment: Option<String>,
}

/// Full API response: meta + list of NAVs
#[derive(Debug, Clone, Deserialize)]
pub struct NavResponse {
    pub meta: NavMeta,
    pub data: Vec<NavData>,
}

/// Date range object (immutable)
#[derive(Debug, Clone, Deserialize)]
pub struct DateRange {
    pub start_date: String,
    pub end_date: String,
}